### PR TITLE
ui: do not write() to channel if builtin pager has terminated

### PR DESCRIPTION
**Pull request by yuja**
# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [ ] I have added tests to cover my changes

### CODE REVIEW

File: cli/src/ui.rs [hunk 1]
```diff
@@ -57,6 +57,11 @@ impl Write for &BuiltinPager {
     }
 
     fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        // Since minus::Pager holds the rx end internally, push_str() doesn't
```
Comment (ilyagr) [inline cli/src/ui.rs:60]: Do you want to add a quick note to the changelog?

**Review by ilyagr — COMMENTED** [review]
Nice! Seems to me like this should work. @ErichDonGubler , WDYT?

In principle, an approach closer to the one from #4187 would give us more control and maybe allow us not to lose output in #4197. I'm not sure it's worth it, especially if we might jump over to `streampager` anyway.

**Review by ilyagr — APPROVED** [review]
I'm hoping a fix to this can go in before the release, but let's wait a day before merging to give people time to look at this.

File: cli/src/ui.rs [hunk 2]
```diff
@@ -57,6 +57,11 @@ impl Write for &BuiltinPager {
...
+        // fail even after the paging thread gets terminated.
+        if self.dynamic_pager_thread.is_finished() {
+            return Err(io::Error::new(io::ErrorKind::BrokenPipe, "pager quit"));
```
Comment (ErichDonGubler) [inline cli/src/ui.rs:63]: suggestion: Unless you want the error message here (I don't think it's used at any point?), you can avoid an allocation for the error message by using `io::Error::from(io::ErrorKind::BrokenPipe)`.

**Review by ErichDonGubler — APPROVED** [review]
Ahh, I like this much better than <https://github.com/martinvonz/jj/pull/4187>. I believe it's more robust, even. Cheers!

File: cli/src/ui.rs [hunk 3]
```diff
@@ -57,6 +57,11 @@ impl Write for &BuiltinPager {
     }
 
     fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        // Since minus::Pager holds the rx end internally, push_str() doesn't
+        // fail even after the paging thread gets terminated.
+        if self.dynamic_pager_thread.is_finished() {
```
Comment (ErichDonGubler) [inline cli/src/ui.rs:62]: praise: I'm feeling silly that I had completely forgotten about `JoinHandle::is_finished`, and hadn't considered this approach in https://github.com/martinvonz/jj/pull/4187. 😆😭

File: cli/src/ui.rs (hunk 2 above)
Comment (yuja) [reply cli/src/ui.rs]: Allocation isn't important, but I like it. Switched to from/into.

**Pull request merged**